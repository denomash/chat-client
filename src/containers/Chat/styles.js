import { makeStyles } from "@mui/styles";

export const useStyles = makeStyles({
  root: {
    display: "flex",
    flexDirection: "column",
    flexGrow: 1,
  },
  messagesContainer: {
    flexGrow: 1,
    overflow: "auto",
  },
  messages: {
    display: "flex",
    flexDirection: "column",
    padding: "0 16px 0 16px",
    justifyContent: "end",
    height: "100%",
  },
  message: {
    padding: 10,
    borderRadius: 4,
  },
  sender: {
    fontSize: 14,
    color: "gray",
  },
  textArea: {
    width: "88%",
    margin: "0 0 6px 6px",
    padding: 16,
    resize: "none",
  },
  button: {
    margin: "0 6px 6px 6px",
    padding: 12,
    resize: "none",
  },
});
