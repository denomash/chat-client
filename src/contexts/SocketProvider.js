import React, { useContext, useEffect, useState } from "react";
import io from "socket.io-client";

const SocketContext = React.createContext();

const useSocket = () => useContext(SocketContext);

const ENDPOINT = "localhost:5000";

const SocketProvider = ({ username, children }) => {
  const [socket, setSocket] = useState();

  useEffect(() => {
    const newSocket = io(ENDPOINT, {
      transports: ["websocket"],
      query: { id: username },
    });
    setSocket(newSocket);

    return () => newSocket.close();
  }, [username]);

  return (
    <SocketContext.Provider value={socket}>{children}</SocketContext.Provider>
  );
};

export { useSocket, SocketProvider };
