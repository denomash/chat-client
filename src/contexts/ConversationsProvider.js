import React, { useCallback, useContext, useEffect, useState } from "react";
import usePersistData from "../hooks/usePersistData";
import { useSocket } from "./SocketProvider";

const ConversationsContext = React.createContext();

const useConversations = () => {
  return useContext(ConversationsContext);
};

const ConversationsProvider = ({ username, children }) => {
  const [conversations, setConversations] = usePersistData("conversations", []);
  const [selectedConversationIndex, setSelectedConversationIndex] = useState(0);

  const socket = useSocket();

  const createConversations = (recipients) => {
    setConversations((prevConversations) => {
      return [...prevConversations, { recipients, messages: [] }];
    });
  };

  const addMessageToConversation = useCallback(
    ({ recipient, recipients, text, sender }) => {
      setConversations((prevConversations) => {
        let convoExists = false;
        let newMessage = { text, sender };
        const newConversations = prevConversations.map((conversation) => {
          const recievers = conversation.recipients.map(
            (recipient) => recipient.username
          );

          if (recievers.includes(recipient)) {
            convoExists = true;
            return {
              ...conversation,
              messages: [...conversation.messages, newMessage],
            };
          }
          return conversation;
        });

        // Check if conversation exists
        // If exists update
        // Otherwise update existing conversations with the new one
        if (convoExists) {
          return newConversations;
        } else {
          // This mostly happens on the new tab we are starting a new conversation with
          return [
            ...prevConversations,
            { recipients, recipient, messages: [newMessage] },
          ];
        }
      });
    },
    [setConversations]
  );

  useEffect(() => {
    if (socket == null) return;

    socket.on("receive-message", addMessageToConversation);
    return () => socket.off("receive-message");
  }, [socket, addMessageToConversation, conversations]);

  const sendMessage = (recipient, recipients, text) => {
    socket.emit("send-message", { recipient, recipients, text });
    addMessageToConversation({ recipients, recipient, text, sender: username });
  };

  const convos = conversations.map((convo, i) => {
    const messages = convo.messages.map((message) => {
      const fromMe = username === message.sender;
      return { ...message, fromMe };
    });

    return {
      ...convo,
      messages,
    };
  });

  return (
    <ConversationsContext.Provider
      value={{
        conversations: convos,
        createConversations,
        selectedConversation: convos[selectedConversationIndex],
        selectConversationIndex: setSelectedConversationIndex,
        sendMessage,
      }}
    >
      {children}
    </ConversationsContext.Provider>
  );
};

export { useConversations, ConversationsProvider };
