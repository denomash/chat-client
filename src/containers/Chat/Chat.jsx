import React, { useCallback, useState } from "react";
import { Box, Button } from "@mui/material";

import { useConversations } from "../../contexts/ConversationsProvider";
import { useStyles } from "./styles";

const Chat = ({ username }) => {
  const [text, setText] = useState("");
  const { sendMessage, selectedConversation } = useConversations();
  const lastMessageRef = useCallback((node) => {
    if (node) {
      node.scrollIntoView();
    }
  }, []);

  const classes = useStyles();

  const onSend = () => {
    const names = selectedConversation.recipients.map(
      (recipient) => recipient.username
    );

    const recipient = names.find((name) => name !== username);

    console.log(username, names, selectedConversation, recipient, { text });
    sendMessage(recipient, names, text);
    setText("");
  };
  return (
    <Box className={classes.root}>
      <Box className={classes.messagesContainer}>
        <Box className={classes.messages}>
          {selectedConversation.messages.map((message, i) => {
            const isLastMessage =
              selectedConversation.messages.length - 1 === i;
            return (
              <Box
                key={i}
                display="flex"
                flexDirection="column"
                alignItems={`${message.fromMe ? "end" : "start"}`}
                padding={1}
                ref={isLastMessage ? lastMessageRef : null}
              >
                <Box
                  className={classes.message}
                  style={{
                    color: `${message.fromMe ? "#fff" : "#333"}`,
                    background: `${message.fromMe ? "#1976d2" : "#cbcbcb"}`,
                    borderRadius: `${message.fromMe ? '10px 0 10px 10px' : "0 10px 10px 10px"}`
                  }}
                >
                  {message.text}
                </Box>
                <Box className={classes.sender}>
                  {message.fromMe ? "You" : message.sender}
                </Box>
              </Box>
            );
          })}
        </Box>
      </Box>

      {/* New text message */}
      <Box display="flex">
        <textarea
          data-testid="chat-input"
          placeholder="Type a message"
          value={text}
          onChange={(e) => setText(e.target.value)}
          className={classes.textArea}
        />
        <Button
          variant="contained"
          className={classes.button}
          disabled={!text}
          onClick={() => onSend()}
        >
          Send
        </Button>
      </Box>
    </Box>
  );
};

export default Chat;
