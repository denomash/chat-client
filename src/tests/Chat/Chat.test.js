import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";

import Chat from "../../containers/Chat/Chat";
import { ConversationsContext } from "../../contexts/ConversationsProvider";
import { SocketProvider } from "../../contexts/SocketProvider";

const mock = {
  selectedConversation: {
    messages: [],
  },
  sendMessage: (recipient, names, text) => {},
};

const Container = () => (
  <SocketProvider username="test">
    <ConversationsContext.Provider value={mock}>
      <Chat username="test" />
    </ConversationsContext.Provider>
  </SocketProvider>
);

it("renders correctly when there are no items", () => {
  const tree = renderer.create(<Container />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render a disabled send button", () => {
  render(<Container />);
  const sendButton = screen.getByRole("button", {
    name: /send/i,
  });
  expect(sendButton).toBeDisabled();
});

it("should render a enabled send button after type", () => {
  render(<Container />);
  const inputEl = screen.getByTestId("chat-input");
  expect(inputEl).toBeInTheDocument();

  fireEvent.change(inputEl, {target: {value: 'test'}})


  const sendButton = screen.getByRole("button", {
    name: /send/i,
  });
  expect(sendButton).toBeEnabled();
});
