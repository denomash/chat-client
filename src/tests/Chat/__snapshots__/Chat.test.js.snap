// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="sc-gsDKAQ makeStyles-root-1 MuiBox-root"
>
  <div
    className="sc-gsDKAQ makeStyles-messagesContainer-2 MuiBox-root"
  >
    <div
      className="sc-gsDKAQ makeStyles-messages-3 MuiBox-root"
    />
  </div>
  <div
    className="sc-gsDKAQ eXKIQs MuiBox-root"
  >
    <textarea
      className="makeStyles-textArea-6"
      data-testid="chat-input"
      onChange={[Function]}
      placeholder="Type a message"
      value=""
    />
    <button
      className="MuiButtonBase-root-KjFju gUleGd MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled MuiButton-root-gygCtr fSFcLn makeStyles-button-7"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Send
    </button>
  </div>
</div>
`;
